File: web_version/pmugg_files/network/vertexNet.js
ms.vertexNet = function() {
	this.halfEdges = [];
	this.group = null;
	this.primal = null;
	this.network = null;

	this.id = ms.counter.add('vertexNet');
};

ms.counter.register('vertexNet');

ms.vertexNet.prototype.getHalfEdges = function() { return this.halfEdges; }
ms.vertexNet.prototype.getPrimal    = function() { return this.primal; }
ms.vertexNet.prototype.getNetwork   = function() { return this.network; }
ms.vertexNet.prototype.getId        = function() { return this.id; }

ms.vertexNet.prototype.setPrimal = function(primal) {
	this.primal = primal;
};

ms.vertexNet.prototype.connectNet = function(network) {
	this.network = network;
	network.addVertex(this);
	return this;
};

ms.vertexNet.prototype.setHalfEdge = function(halfEdge, index) {
	this.halfEdges[index] = halfEdge;
};

ms.vertexNet.prototype.copyConnection = function(copy) {
	var copyNet = copy.getNetwork();
	var self = this;
	this.halfEdges = copy.getHalfEdges().map((halfEdge) => {
		return self.network.convertHalfEdge(copyNet, halfEdge);
	});
	var copyGroup = copy.getGroup();
	this.group = copyGroup ?
		this.network.convertConnectorGroup(copyNet, copyGroup) :
		null;
};

ms.vertexNet.prototype.export = function() {
	var self = this;
	var group = this.getGroup();
	return {
		halfEdges: this.getHalfEdges().map((halfEdge) => {
			return self.network.halfEdgeIndex(halfEdge);
		}),
		group: group ? this.network.connectorGroupIndex(group) : -1,
	}
};

ms.vertexNet.prototype.import = function(json) {
	var self = this;
	this.halfEdges = json.halfEdges.map((halfEdge) => {
		return self.network.getHalfEdges()[halfEdge];
	});
	this.group = json.group >= 0 ?
		this.network.getConnectorGroups()[json.group] :
		null;
};

// This is just a sanity check. This should be true if this is valid.
ms.vertexNet.prototype.inNetwork = function() {
	return this.network.getVertices().includes(this);
};

ms.vertexNet.prototype.compare = function(endpointA, endpointB) {
	return ms.vertexType.compare(endpointA, endpointB);
};

ms.vertexNet.prototype.getGroup = function() {
	return this.group;
};

ms.vertexNet.prototype.setGroup = function(group) {
	this.group = group;
};

ms.vertexNet.prototype.highlight = function(view) {
	var options = ms.halfEdgeSet.create(this.halfEdges);
	this.network.highlight(view, options);
};

ms.vertexNet.prototype.print = function() {
	var interior = this.primal.getInterior();
	if (interior != this) {
		console.log ('Boundary Vertex');
		var halfEdge = interior.getHalfEdges()[0];
		if (halfEdge.getEdge()) {
			halfEdge.getEdge().print();
		} else {
			halfEdge.getPrev().getEdge().print();
		}		
	} else {
		ms.highlight(this);
	}
};

ms.vertexNet.prototype.requiresShapeView = function () {
	return true;
};

File: cpp_version/pmugg/network/vertex_net.cpp
#include "pch.h"
#include "vertex_net.h"
#include "network.h"
#include "connector_group.h"
#include "half_edge_net.h"
#include "../util/util.h"

namespace ms {

int VertexNet::nextId = 0;

VertexNet::VertexNet()
    : network(nullptr)
    , type(nullptr)
    , id(nextId++) {}

VertexNet* VertexNet::connectNet(Network* net) {
    network = net;
    network->addVertex(this);
    return this;
}

void VertexNet::setHalfEdge(HalfEdgeNet* halfEdge, int index) {
    if (index >= halfEdges.size()) {
        halfEdges.resize(index + 1, nullptr);
    }
    halfEdges[index] = halfEdge;
}

int VertexNet::connectorIndex() const {
    auto bVertices = network->getBVertices();
    auto it = std::find(bVertices.begin(), bVertices.end(), this);
    if (it != bVertices.end()) {
        // Compute the index
        return std::distance(bVertices.begin(), it);
    } else {
        return -1;
    }
}

void VertexNet::copyConnection(const VertexNet* copy) {
    auto* copyNet = copy->getNetwork();

    halfEdges.clear();
    auto copyHalfEdges = copy->getHalfEdges();
    halfEdges.reserve(copyHalfEdges.size());
    
    for (auto* halfEdge : copyHalfEdges) {
        halfEdges.push_back(halfEdge ? network->convertHalfEdge(copyNet, halfEdge) : nullptr);
    }
}

bool VertexNet::inNetwork() const {
    return network && std::find(network->getVertices().begin(),
                              network->getVertices().end(),
                              this) != network->getVertices().end();
}

EdgeNet* VertexNet::interiorEdge() const {
    for (auto* halfEdge : halfEdges) {
        if (halfEdge) {
            auto edge = halfEdge->getEdge();
            if (edge) {
                return edge;
            }
        }
    }
    return nullptr;
}

void VertexNet::import(const Json& json) {
    halfEdges.clear();
    auto& networkHalfEdges = network->getHalfEdges();
    
    if (json.contains("halfEdges")) {
        for (const auto& index : json["halfEdges"]) {
            int idx = index.get<int>();
            halfEdges.push_back(idx >= 0 ? networkHalfEdges[idx] : nullptr);
        }
    }
}

} // namespace ms 